File: patterns/singleton.js
class ConfigManager {
  constructor() {
    if (ConfigManager.instance) {
      return ConfigManager.instance;
    }

    this.settings = { theme: "light" };
    ConfigManager.instance = this;
    return this;
  }

  set(key, value) {
    this.settings[key] = value;
  }

  get(key) {
    return this.settings[key];
  }
}

ConfigManager.instance = null;

if (require.main === module) {
  const cfg1 = new ConfigManager();
  const cfg2 = new ConfigManager();
  cfg1.set("language", "en");
  console.log(cfg2.get("language"));
}

module.exports = { ConfigManager };

File: patterns/singleton.py
"""Singleton example: configuration manager."""

class ConfigManager:
    _instance = None

    def __new__(cls):
        # __new__ guarantees only one instance is created.
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance.settings = {"theme": "light"}
        return cls._instance

    def set(self, key, value):
        self.settings[key] = value

    def get(self, key):
        return self.settings.get(key)


if __name__ == "__main__":
    cfg1 = ConfigManager()
    cfg2 = ConfigManager()
    cfg1.set("language", "en")
    print(cfg2.get("language"))  # Prints "en", proving both references share state.
